lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+            #     contexts).  These messages are fed into
+            #     `Endpoint.broadcast`
+            #
+            nursery.start_soon(
+                self._process_broadcasts, broadcast_receive_channel, nursery
+            )
+
+            #
+            # _process_received_messages:
+            #     Manages a channel which all incoming
+            #     event broadcasts are placed on, running each event through
+            #     the internal `_process_item` handler which handles all of the
+            #     internal logic for the various
+            #     `request/response/subscribe/stream/wait_for` API.
+            #
+            nursery.start_soon(
+                self._process_received_messages, message_receive_channel, nursery
+            )
+
+            #
+            # _process_connections:
+            #     When `Endpoint.connect_to_endpoint` is called, the actual
+            #     connection process is done asynchronously by putting the
+            #     `ConnectionConfig` onto a queue which this process
+            #     retrieves to establish the new connection.  This includes
+            #     handing the `RemoteEndpoint` oblect off to the `RemoteManager`
+            #     which takes care of the connection lifecycle.
+            #
+            nursery.start_soon(
+                self._process_connections, connection_receive_channel, nursery
+            )
+
+            #
+            # _monitor_subscription_changes
+            #    Monitors an event for local changes to subscriptions to
+            #    propagate those changes to remotes.
+            nursery.start_soon(self._monitor_subscription_changes)
+
+            # wait for the sub-processes to start before marking the endpoint
+            # as running.
+            await self._message_processing_loop_running.wait()
+            await self._connection_loop_running.wait()
+            await self._process_broadcasts_running.wait()
+
+            # marke the endpoint as running.
+            self._running.set()
+
+            await self.wait_stopped()
+
+            nursery.cancel_scope.cancel()
+
+    async def _cleanup(self) -> None:
+        # Cleanup stateful things.
+        await self._message_send_channel.aclose()
+        await self._connection_send_channel.aclose()
+        await self._broadcast_send_channel.aclose()
+
+        del self._message_send_channel
+        del self._connection_send_channel
+        del self._broadcast_send_channel
+
+        self.logger.debug("%s: stopped", self)
+
+    #
+    # Background processes
+    #
+    async def _process_broadcasts(
+        self,
+        channel: trio.abc.ReceiveChannel[ConcreteBroadcast],
+        nursery: trio_typing.Nursery,
+    ) -> None:
+        self._process_broadcasts_running.set()
+        async for (item, config) in channel:
+            nursery.start_soon(self.broadcast, item, config)
+
+    async def _process_connections(
```

Comment:
Why did you decide to pull this out into another coro+channel? If the nursery is added somewhere like `self.nursery` then `connect_to_endpoint` could do the connect and then call `nursery.start_soon`.

Reply:
This might be a leftover from a previous design problem that no-longer exists. Will very and remove if it isn't needed.

Reply:
Ah, because it needs access to the nursery to run the endpoint in the background and I much prefer trying to keep access to the nursery isolated so this indirection is the cost.